arraycontext/container/traversal.py

```diff
@@ -624,17 +626,40 @@ def _flatten(subary: ArrayOrContainerT) -> None:
                         "This functionality needs to be implemented by the "
                         "array context.") from exc
 
-            result.append(flat_subary)
+            result = [flat_subary]
         else:
+            result = []
             for _, isubary in iterable:
-                _flatten(isubary)
+                result.extend(_flatten(isubary))
 
-    _flatten(ary)
+        return result
 
-    if len(result) == 1:
-        return result[0]
+    def _flatten_without_leaf_class(subary: ArrayOrContainerT) -> Any:
+        result = _flatten(subary)
+
+        if len(result) == 1:
+            return result[0]
+        else:
+            return actx.np.concatenate(result)
+
+    def _flatten_with_leaf_class(subary: ArrayOrContainerT) -> Any:
+        if type(subary) is leaf_class:
```

Comment:
Exact match or `isinstance`? What's the precedent from other `leaf_class`es? Also: document.

Reply:
All the other `rec_map` things don't use `isinstance` either. I'm not sure which choice would be better, since we mostly just have `DOFArray`s at the moment..

Reply:
Mentioned in 5e87623 that subclasses are not considered.